src/libraries/Microsoft.Extensions.Options/gen/Generator.cs

```diff
@@ -0,0 +1,53 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.Extensions.Options.Generators
+{
+    [Generator]
+    public class Generator : IIncrementalGenerator
+    {
+        public void Initialize(IncrementalGeneratorInitializationContext context)
+        {
+            IncrementalValuesProvider<TypeDeclarationSyntax> typeDeclarations = context.SyntaxProvider
+                .ForAttributeWithMetadataName(
+                    SymbolLoader.OptionsValidatorAttribute,
+                    (node, _) => node is TypeDeclarationSyntax,
+                    (context, _) => context.TargetNode as TypeDeclarationSyntax)
+                .Where(static m => m is not null);
+
+            IncrementalValueProvider<(Compilation, ImmutableArray<TypeDeclarationSyntax>)> compilationAndTypes =
+                context.CompilationProvider.Combine(typeDeclarations.Collect());
+
+            context.RegisterSourceOutput(compilationAndTypes, static (spc, source) => HandleAnnotatedTypes(source.Item1, source.Item2, spc));
+        }
+
+        private static void HandleAnnotatedTypes(Compilation compilation, ImmutableArray<TypeDeclarationSyntax> types, SourceProductionContext context)
+        {
+            if (!SymbolLoader.TryLoad(compilation, out var symbolHolder))
+            {
+                // Not eligible compilation
+                return;
+            }
+
+            var parser = new Parser(compilation, context.ReportDiagnostic, symbolHolder!, context.CancellationToken);
+
+            var validatorTypes = parser.GetValidatorTypes(types);
```

Comment:
As a next step, you might want to move the parser logic inside an `IncrementalValuesProvider.Select` declaration so that you end up with an `IncrementalValuesProvider<ValidatorType>`. 

To achieve this, the Parser class needs to be updated so that it just stores `Diagnostic` values, to be reported by the `SourceProductionContext` later:

https://github.com/dotnet/runtime/blob/be6a1389c852fdc96a919769867fb3f3ddb0fbde/src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs#L47-L57